src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -3554,6 +3554,36 @@ std::optional<UInt64> StorageReplicatedMergeTree::totalRows() const
     return res;
 }
 
+std::optional<UInt64> StorageReplicatedMergeTree::totalRowsByPartitionPredicate(const SelectQueryInfo & query_info, const Context & context) const
+{
+    auto metadata_snapshot = getInMemoryMetadataPtr();
+    const auto & partition_key = metadata_snapshot->getPartitionKey();
+    Names partition_key_columns = partition_key.column_names;
+    KeyCondition key_condition(query_info, context, partition_key_columns, partition_key.expression, true /* exact_match */);
+    if (key_condition.alwaysUnknownOrTrue())
+        return {};
+    std::unordered_map<String, bool> partition_filter_map;
+    size_t res = 0;
+    foreachCommittedParts([&](auto & part)
```

Comment:
We can analyze partitions first by simply evaluating the condition (with VirtualColumnUtils) on partition expression - to count or filter out whole partitions before filtering parts.

Reply:
I was wrong.